cmd/snap-bootstrap/initramfs_systemd_mount.go

```diff
@@ -190,56 +254,40 @@ func doSystemdMountImpl(what, where string, opts *systemdMountOptions) error {
 	if opts.Private {
 		options = append(options, "private")
 	}
-	if opts.Overlayfs {
-		if len(opts.LowerDirs) <= 0 || len(opts.UpperDir) <= 0 || len(opts.WorkDir) <= 0 {
-			return fmt.Errorf("cannot mount %q at %q: missing arguments for overlayfs mount. lowerdir, upperdir, workdir are needed.", what, where)
-		}
-
-		if strings.ContainsAny(opts.UpperDir, forbiddenChars) {
-			return fmt.Errorf("cannot mount %q at %q: upperdir overlayfs mount option contains forbidden characters. %q contains one of %q.", what, where, opts.UpperDir, forbiddenChars)
-		}
-		if strings.ContainsAny(opts.WorkDir, forbiddenChars) {
-			return fmt.Errorf("cannot mount %q at %q: workdir overlayfs mount option contains forbidden characters. %q contains one of %q.", what, where, opts.WorkDir, forbiddenChars)
-		}
+	if opts.OverlayFsOpts != nil {
+		args = append(args, "--type=overlay")
 
-		var lowerDirs strings.Builder
-		for i, d := range opts.LowerDirs {
-			if strings.ContainsAny(d, forbiddenChars) {
-				return fmt.Errorf("cannot mount %q at %q: lowerdir overlayfs mount option contains forbidden characters. %q contains one of %q.", what, where, d, forbiddenChars)
-			}
+		o := opts.OverlayFsOpts
 
-			// This is used for splitting multiple lowerdirs as done in
-			// https://elixir.bootlin.com/linux/v6.10.9/C/ident/ovl_parse_param_split_lowerdirs
-			if i != 0 {
-				lowerDirs.WriteRune(':')
-			}
+		err := o.Validate()
+		if err != nil {
+			return fmt.Errorf("cannot mount %q at %q: %w", what, where, err)
+		}
 
-			lowerDirs.WriteString(d)
+		lowerDirs, err := o.ValidateLowerDirs()
+		if err != nil {
+			return fmt.Errorf("cannot mount %q at %q: %w", what, where, err)
 		}
-		options = append(options, fmt.Sprintf("lowerdir=%s", lowerDirs.String()))
-		options = append(options, fmt.Sprintf("upperdir=%s", opts.UpperDir))
-		options = append(options, fmt.Sprintf("workdir=%s", opts.WorkDir))
-	}
-	if opts.VerityHashDevice != "" && opts.VerityRootHash == "" {
-		return fmt.Errorf("cannot mount %q at %q: mount with dm-verity was requested but a root hash was not specified", what, where)
-	}
-	if opts.VerityRootHash != "" && opts.VerityHashDevice == "" {
-		return fmt.Errorf("cannot mount %q at %q: mount with dm-verity was requested but a hash device was not specified", what, where)
-	}
 
-	if strings.ContainsAny(opts.VerityHashDevice, forbiddenChars) {
-		return fmt.Errorf("cannot mount %q at %q: dm-verity hash device path contains forbidden characters. %q contains one of %q.", what, where, opts.VerityHashDevice, forbiddenChars)
+		options = append(options, fmt.Sprintf("lowerdir=%s", lowerDirs))
+		options = append(options, fmt.Sprintf("upperdir=%s", o.UpperDir))
+		options = append(options, fmt.Sprintf("workdir=%s", o.WorkDir))
```

Comment:
IMO it would be much nicer to have a method on the mount options, to return a slice of arguments for mount and an error. All the logic of how things are composed would live in those functions then. I strongly believe you can then make this an interface and only hold one value.

Reply:
I am not sure if you also meant to hide the validation logic behind these functions but I did and I think it looks better. Let me know what you think.

Reply:
Yes that was it, thank you :)